keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1179 @@
+  will be addressed separate from this KEP
+
+## Proposal
+
+### Component Flags
+
+Kubernetes components (apiservers, controller managers, schedulers) will offer a
+`--compatibility-version` flag that can be set to any of the previous three
+minor versions. If unset, the compatibility version defaults to the `<major.minor>`
+version of the binary version.
+
+### Changes to Feature Gates
+
+Features will track version information, i.e.:
+
+```go
+map[Feature]VersionedSpecs{
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("1.27"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("1.28"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Alpha},
+		},
+		featureC: VersionedSpecs{
+			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Beta},
+		},
+		featureD: VersionedSpecs{
+			{Version: mustParseVersion("1.26"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.28"), Default: true, PreRelease: Deprecated},
+		}
+```
+
+When a component starts, feature gates will be compared against the
+compatibility version to determine which features to enable to match the set of
+features that where enabled for the Kubernetes version the compatibility version
+is set to.
+
+#### Feature Gate Lifecycles
+
+`--feature-gates` must behave the same as it did for the Kubernetes
+version the compatibility version is set to. I.e. it must be possible to use
+`--feature-gates` to disable features that were beta, and enable feature that
+were alpha in the Kubernetes version the compatibility version is set to. One
+important implication of this requirement is that feature gating must be kept in
+the Kubenetes codebase until a feature has reached GA (or been removed) for N-3
+releases.
+
+For example, a feature that is promoted once per release would look something like:
+
+```go
+map[Feature]VersionedSpecs{
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("1.26"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.27"), Default: true, PreRelease: Beta},
+			{Version: mustParseVersion("1.28"), Default: true, PreRelease: GA},
+		},
+}
+```
+
+The lifecycle of the feature would be:
+
+| Release | Stage | Feature tracking information                      |
+| ------- | ----- | ------------------------------------------------- |
+| 1.26    | alpha | Alpha: 1.26                                       |
+| 1.27    | beta  | Alpha: 1.26, Beta: 1.27 (on-by-default)           |
+| 1.28    | GA    | Alpha: 1.26, Beta: 1.27 (on-by-default), GA: 1.28 |
+| 1.29    | GA    | Alpha: 1.26, Beta: 1.27 (on-by-default), GA: 1.28 |
+| 1.30    | GA    | Alpha: 1.26, Beta: 1.27 (on-by-default), GA: 1.28 |
+| 1.31    | GA    | **Feature implementation becomes part of normal code, Feature gate and feature tracking information may be removed from code** |
```

Comment:
Maybe clarify that it's only the moment when you can remove the "if featureGate enabled { // implement feature }" - as this is what is changing compared to now.

[I know that it is supposed to say so, but let's maybe make it more explicit.]

Reply:
Thx. Clarified